BankManagerConsoleApp/BankManagerTest/LinkedListTest.swift

```diff
@@ -0,0 +1,71 @@
+//
+//  LinkedListTest.swift
+//  LinkedListTest
+//
+//  Created by Eddy, dudu on 2022/04/25.
+//
+
+import XCTest
+
+class LinkedListTest: XCTestCase {
+
+    var sut: LinkedList<Int>!
+
+    override func setUpWithError() throws {
+        sut = LinkedList<Int>()
+    }
+
+    override func tearDownWithError() throws {
+        sut = nil
+    }
```

Comment:
`setUpWithError`와 `tearDownWithError`은 어느 시점에 실행되고 어떤 상황에 사용해야 할까요? 👀 

`XCTest` 프레임워크에서는 각 테스트 케이스를 실행할 때, 하나의 테스트 케이스당 하나의 테스트 클래스 인스턴스(`LinkedListTest`)를 각각 생성해서 실행하게 됩니다.

지금 구현에선 
```suggestion
    let sut: LinkedList<Int> = LinkedList<Int>()
```
요렇게 선언해도 의도하신대로 동작할 거예요.

앱 전역에서 상태를 공유하는 싱글톤 객체나 데이터베이스 등과 의존성을 가지는 테스트 케이스를 작성할 때, 각 테스트 케이스에서의 수정이 다른 테스트 케이스의 결과에 영향을 줄 수 있으니 `setUpWithError`와 `tearDownWithError` 내부에 테스트를 위한 설정, 초기화 코드를 작성해줘야 하고 지금은 굳이 사용할 필요 없어보이네요!

Reply:
저희가 처음 생각했을 때 `setUpWithError`와 `tearDownWithError` 이 두 메서드는 테스트 코드 내에서 공유되어 sut이 사용되고 있기에 테스트 코드 작성시 마다 이 sut이 초기화가 스스로 되지 않아서 매번 테스트 메서드를 실행하기전에 위의 두 메서드를 통해 초기화 해줘야한다고 생각했습니다. 하지만 또치가 해주신 방법으로 해보니 테스트코드가 다 잘 돌아갔으며 매번 끝난 후 sut이 empty가 되는 것을 확인했습니다.
그러면 실제 `setUpWithError`와 `tearDownWithError` 이 메서드는 테스트 코드에서의 공유된 자원이 아닌, 실제 코드에서 공유된 자원이 있을때만 초기화를 해줘야하는 메서드인건가요? 

```swift
let sut: LinkedList<Int> = LinkedList<Int>()
```
이 코드는 매 테스트 메서드가 실행되기 전마다 초기화가 되는게 맞는건가요?

Reply:
> 그러면 실제 setUpWithError와 tearDownWithError 이 메서드는 테스트 코드에서의 공유된 자원이 아닌, 실제 코드에서 공유된 자원이 있을때만 초기화를 해줘야하는 메서드인건가요?

실제 코드에서 공유된 자원일 수도 있긴 하지만, 테스트 케이스의 로직이 실제 코드에 영향을 주면 안되기에 다른 예시를 들어야 할 것 같아요 🤔 

예를 들어 `UserDefaults`와 의존성을 가지고 외부로부터 `UserDefaults.standard`를 주입받아 로컬에 값을 저장하는 로직을 가지고 있는 객체가 있어요.
이 객체를 테스트할 땐, 실제 앱의 로직에 영향을 주면 안되기에 `standard`에 접근하지 않고 `init(suiteName:)`으로 테스트에서만 사용할 `UserDefaults`를 만들어서 주입해줄거에요.
그러면 데이터가 저장되는 곳이 분리되어 실제 앱 로직엔 영향을 주지않지만, 어찌됐건 데이터를 저장은 하기 때문에 전역에서 공유하게되어 각 테스트 케이스의 로직이 서로 영향을 주는 형태가 됩니다.
이 때 각각의 테스트 케이스가 실행되기 전, 후 상태를 초기화해주기위해 `setUpWithError`와 `tearDownWithError`를 사용하게 될거에요.

> let sut: LinkedList<Int> = LinkedList<Int>()
이 코드는 매 테스트 메서드가 실행되기 전마다 초기화가 되는게 맞는건가요?

넵 요거 확인해보시려면 `LinkedList` 생성 시점에 로그를 찍어 테스트 케이스 개수만큼 찍히는지 확인할 수 있겠네요.
약간 깊게 들어가면 https://github.com/apple/swift-corelibs-xctest/blob/main/Sources/XCTest/Private/XCTestCaseSuite.swift#L25-L28 케이스마다 클래스의 인스턴스를 생성해주는 것을 코드로도 확인할 수 있습니다!